docs/guides/implementation-guides/how-to-implement-dependency-injection.md

```diff
@@ -3,13 +3,144 @@ id: how-to-implement-dependency-injection
+[Dependency injection (DI)](https://en.wikipedia.org/wiki/Dependency_injection) allows developers to write cleaner, more modular, and testable code. It accomplishes this by creating discrete services that are passed around/created as needed. It is especially useful when you are developing multi-platform applications that need specific service implementations for each platform.
 
 This guide will show you how to use Dependency Injection (DI) with _Avalonia UI_ and the MVVM pattern. 
 
-<img src={MvvmArchitectureDiagram} alt=""/>
+Let's assume that your MainViewModel has a dependency on IMyService. The constructor for ViewModel should looks like this:
+
+```csharp
+private readonly IMyService _myService;
+
+public MainViewModel(IMyService myService)
+{
+    _myService = myService;
+}
+```
+
+You could solve that by instantiating `MyService` and passing as a parameter everytime you need a `MainViewModel` instance, like this:
+
+```csharp
+var window = new MainWindow
+{
+    // Note: we assume that MyService implements IMyService
+    DataContext = new MainViewModel(new MyService())
+}
+```
+
+But this will get quickly difficult as when your application scale and your classes have more dependencies, hence the need of DI.
+
+These are the steps you need to do in order to resolve that dependencu using DI.
+
+## Step 1: Install the NuGet package for DI
+There are many dependency injection (DI) container providers available ([DryIoC](https://github.com/dadhi/DryIoc), [Autofac](https://github.com/autofac/Autofac), [Pure.DI](https://github.com/DevTeam/Pure.DI)) but this guide will only focus on `Microsoft.Extensions.DependencyInjection`. This is a lightweight, extensible dependency injection container that is part of the .NET Framework. It provides an easy-to-use and convention-based way to add DI to .NET applications, including Avalonia-based desktop applications.
+
+```shell
+dotnet add package Microsoft.Extensions.DependencyInjection
+```
+
+## Step 2: Add ServiceCollectionExtensions 
+The following code is creating an extension for IServiceCollection. This is where you will register all your platform independent shared dependencies. It's recomended to create that class in the shared project used by all avalonia target platform.
+
+```csharp
+public static class ServiceCollectionExtensions {
+    public static void AddCommonServices(this IServiceCollection collection) {
+        collection.AddTransient<MainViewModel>();
+        collection.AddTransient<MyService>();
+    }
+}
+```
+
+## Step 3: Add ServiceManager
+The following code creates a ServiceManager class that is used to retrieve an instance of a given object. It's recomended to create that class in the shared project used by all avalonia target platform.
+
+```csharp
+public static class ServiceManager {
+    public static IServiceProvider Locator { get; private set; }
+
+    public static void InitialiseLocatorFromCollection(IServiceCollection collection) {
+        if (Locator != null)
```

Comment:
I would expect this to throw an exception instead of just silently failing if the locator can't be set.

Reply:
indeed it's working on my side, but it should not...

Reply:
It wont throw an exception here since you are ignoring it when the condition is false.